projects/bb3/agents/opt_bb3_agent.py

```diff
@@ -285,27 +274,65 @@ def get_orm_observation(
         """
         agent = self.agents[Module.OPENING_DIALOGUE]
         agent.reset()
-        for i, mem in enumerate(opening_memories):
+        prefixed_memories = {}
+        for mem, val in opening_memories.items():
             mem = MemoryUtils.maybe_add_memory_prefix(mem, 'partner', self.MODEL_TYPE)
-            opening_memories[i] = mem
+            prefixed_memories[mem] = val
 
         new_obs = copy.deepcopy(observation)
-        new_obs.force_set(
-            'text', self._check_and_limit_len('\n'.join(opening_memories))
+        memories_to_use = MemoryUtils.get_available_memories(
+            '',
+            prefixed_memories,
+            set(),
+            dictionary=self.dictionary,
+            **self._get_memory_heuristic_values(),
         )
-        new_obs.force_set('memories', opening_memories)
+        if not memories_to_use:
+            # we need at least one memory to open with...
+            memories_to_use = random.sample(list(prefixed_memories.keys()), 1)
+
+        new_obs.force_set('text', self._check_and_limit_len('\n'.join(memories_to_use)))
```

Comment:
Why are we setting `text` here?

Reply:
we're rebuilding the observation for the opening model; `'text'` is what the model is seeing